### ISSUE TITLE
Bug on Ubuntu 14.10

### ISSUE THREAD

**Issue by ceholden** [ceholden/segment#1]
Segmentation works without issue on BU cluster (CentOS 6) and produces the exact answer the original compilation of `segment` produces. The program runs on Ubuntu 14.10, but there is a bug in the code that results in close to zero merges. After spending a lot of time in `gdb`, I've narrowed it to an unusual behavior of [one variable in `reg_nnbr` within `src/region.c`](https://github.com/ceholden/segment/blob/master/segment/src/region.c#L731).

``` C
begin_set(Nset);
    {
        float           mdist2 = MAXFLOAT;
        float           ndist2;
        REGION_ID       nnbr = 0;

        while (get_from_set(Nset, (addr_t) & nbr)) {
            ndist2 = reg_dist2(Spr->nbands,
                               regid_to_Ctr(Spr, rid),
                               regid_to_Ctr(Spr, nbr));
            if (ndist2 > mdist2)
                continue;
            else if (ndist2 < mdist2) {
                mdist2 = ndist2;
                nnbr = nbr;
            } else {

                /*
                 * This is biased, but it does give some randomness to nnbr
                 * selection.
                 */
                if (flip()) {
                    nnbr = nbr;
                }
            }
        }
```

`mdist2` changes from `MAXFLOAT` to `0` when the code enters the `while` loop. No idea why it happens. 

`gcc` information for CentOS and Ubuntu:

CentOS:

``` bash
$ gcc -v
Using built-in specs.
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-languages=c,c++,objc,obj-c++,java,fortran,ada --enable-java-awt=gtk --disable-dssi --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-1.5.0.0/jre --enable-libgcj-multifile --enable-java-maintainer-mode --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --disable-libjava-multilib --with-ppl --with-cloog --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.4.7 20120313 (Red Hat 4.4.7-3) (GCC)
```

Ubuntu 14.10:

``` bash
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.9/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 4.9.1-16ubuntu6' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.9.1 (Ubuntu 4.9.1-16ubuntu6)
```

Also not sure if it's caused by a `gcc` version, but this is the only thing I can pick up on.

**Comment by ceholden** [comment]
Note that if you change:

``` C
float           mdist2 = MAXFLOAT;
```

to

``` C
double           mdist2 = MAXFLOAT;
```

the value of `mdist2` does not change when you enter the `while` loop. However, even if you recast the value of `mdist2` back to a `float` before assigning it to `regid_to_nnbr(Spr, rid).nbr_d2`, changing `mdist2` to a `double` causes a segfault later in the code.

**Comment by ceholden** [comment]
Fixed in ecaff4914f9b0553ea6ccb03ba3dfcf7c5baed98

Turns out difference wasn't with `gcc` but memory differences on the cluster and my laptop. Because the cluster nodes have so much memory and many users, the memory location of `nbr` used in `get_from_set` was far away from the memory location of `mdist2` due to random chance. On my laptop, the memory was contiguous and so the assignment as an 8-byte `long` overwrote the value of `mdist2`.

Thanks to Robert Putnam from BU IS&T Research Computing for help debugging the issue.
